hazelcast-client/src/test/java/com/hazelcast/client/cache/ClientJCacheTypeChecks.java

```diff
@@ -0,0 +1,373 @@
+        cachingProvider3.close();
+        HazelcastClient.shutdownAll();
+        Hazelcast.shutdownAll();
+    }
+
+    @Test(expected = ClassCastException.class)
+    public void test_check_types_on_put() throws Exception {
+        final String cacheName = randomMapName();
+
+        CacheManager cacheManager = cachingProvider3.getCacheManager();
+        assertNotNull(cacheManager);
+
+        assertNull(cacheManager.getCache(cacheName));
+
+        CompleteConfiguration<Integer, String> config = new CacheConfig<Integer, String>()
+                .setTypes(Integer.class, String.class);
+
+        Cache<Integer, String> cache = cacheManager.createCache(cacheName, config);
+        assertNotNull(cache);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                CacheManager cm1 = cachingProvider1.getCacheManager();
+                assertNotNull(cm1.getCache(cacheName, Integer.class, String.class));
+                CacheManager cm2 = cachingProvider2.getCacheManager();
+                assertNotNull(cm2.getCache(cacheName, Integer.class, String.class));
+            }
+        });
+
+        Cache rawCache = (Cache) cache;
+        rawCache.put(new Object(), new Object());
+    }
+
+    @Test(expected = ClassCastException.class)
+    public void test_check_types_on_get() throws Exception {
+        final String cacheName = randomMapName();
+
+        CacheManager cacheManager = cachingProvider3.getCacheManager();
+        assertNotNull(cacheManager);
+
+        assertNull(cacheManager.getCache(cacheName));
+
+        CompleteConfiguration<Integer, String> config = new CacheConfig<Integer, String>();
+
+        Cache<Integer, String> cache = cacheManager.createCache(cacheName, config);
+        assertNotNull(cache);
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                CacheManager cm1 = cachingProvider1.getCacheManager();
+                assertNotNull(cm1.getCache(cacheName));
+                CacheManager cm2 = cachingProvider2.getCacheManager();
+                assertNotNull(cm2.getCache(cacheName));
+            }
+        });
+
+        Cache rawCache = (Cache) cache;
+        rawCache.put(1, 1);
+
+        Class<?> clazz = Class.forName("com.hazelcast.client.cache.impl.AbstractClientCacheProxyBase");
+        Field configField = clazz.getDeclaredField("cacheConfig");
+        configField.setAccessible(true);
+
+        CacheConfig cacheConfig = (CacheConfig) configField.get(rawCache);
+        cacheConfig.setTypes(Integer.class, String.class);
+
+        cache.get(1);
+    }
+
+    @Test(expected = ClassCastException.class)
+    public void test_check_types_on_getandremove() throws Exception {
+        final String cacheName = randomMapName();
+
+        CacheManager cacheManager = cachingProvider3.getCacheManager();
+        assertNotNull(cacheManager);
```

Comment:
why do you check cacheManager in every test? As far as i learned from @pveentjer , if you don't sure about cacheManager you must create another test for it.